### ISSUE TITLE
Process for promoting images to "stable" tag

### ISSUE THREAD

**Issue by joonas-fi** [tiiuae/fog-ros-baseimage#1]
tl;dr: go to section "Promoting to stable: more formal process"

Background
---------------

In the start from concrete projects we referred to the base image with the explicit version tag: `FROM ghcr.io/tiiuae/fog-ros-baseimage:sha-7857f7d`.

Pros: version we get is always the same. Nothing breaks unexpectedly.
Con: when we release new base image version, we need to change the reference in all concrete repositories.

`FROM ghcr.io/tiiuae/fog-ros-baseimage:latest` would not be good, because it's a fast-moving target and prone to breaks: latest usually means the latest built image from latest commit to the main branch, which doesn't have much stability guarantees.

We had a small brainstorming session and ended up with suggestion of using `FROM ghcr.io/tiiuae/fog-ros-baseimage:stable`, where `stable` is conceptually similar to the `latest` tag in the sense that:

- it's "special"
- it's dynamic: the concrete image version it points to, changes over time (now `stable` points to v1, tomorrow `stable` will point to v2)

This seems like a good middle ground between the riskiness of `latest` and the maintenance burden of explicit version pointers.

Promoting to stable
------------------------

We can implement `stable` tag right now by this process: when we want to promote, someone runs this on their computer:

```console
$ docker pull ghcr.io/tiiuae/fog-ros-baseimage:sha-7857f7d
$ docker tag ghcr.io/tiiuae/fog-ros-baseimage:sha-7857f7d ghcr.io/tiiuae/fog-ros-baseimage:stable
$ docker push ghcr.io/tiiuae/fog-ros-baseimage:stable
```

Promoting to stable: more formal process
-----------------------------------

----> this is what this issue is about <-----

@mehmetkillioglu suggested that we could have a separate branch in Git for stable, and merging to it could be restricted to PRs + code review approvals. I like the idea that we could get a clear paper trail for what was promoted to stable, when and why.

The concrete implementation needs some figuring out:

- Ideally we would promote existing container image to stable.
	* It would be risky to promote code as the primary artifact to stable, because that would mean that the "stable" image gets rebuilt from the code (= Dockerfile) and we have to just hope and pray that the resulting "stable" is functionally equivalent to the image candidate that we hopefully tested first as being a good candidate for promoting.
- If we're ideally promoting an existing image to stable (like you see in the `$ docker` pull+tag+push), having a separate stable branch just for process reasons (one can code review the difference of last and new proposed stable) is in some sense redundant. Though this might be a non-issue.
- If the automation needs to be able to detect `ghcr.io/tiiuae/fog-ros-baseimage:sha-7857f7d` as the tag-to-promote, how do we resolve it from GitHub action workflow? I guess it's resolvable from merge commit by looking at the parent revision that comes from the main branch. So we'd probably have a separate workflow that triggers only on stable branch merges, that will move the stable pointer?

Other considerations
---------------------------

I was rambling to @maseabunikie  earlier about a script we could write that we would run before each version release to check that all containers use same base image version.

this "stable" change makes the problem a bit harder. currently this script would be easy to write:
1. download Dockerfile  of each concrete repo
1. check that the string `FROM ghcr.io/tiiuae/fog-ros-baseimage:sha-7857f7d` is present in all Dockerfiles

now if we move to using "indirect" image references (indirect = the actual image of "stable" changes with time) and all Dockerfiles contain the same string ":stable", it's harder to work out if all projects have the same image version.

but that's not a big issue, we can make the script not look at the Dockerfile but the container image instead: pull each image and from the layer metadata ensure that each concrete image uses the same base image version

**Comment by mehmetkillioglu** [comment]
For further automation, using [repository-dispatch](https://github.com/marketplace/actions/repository-dispatch) plugin for the automated triggers to child repositories that are using this `fog-ros-baseimage`. The idea is to automatically update the base image in the containers when the `stable` version is updated (or PR merged to `stable` branch.

On the child repositories that uses base image, the action yaml have this along side with the push;

```
on:
  repository_dispatch:
    types: [base-image-update]
  push:
```

And in the `fog-ros-baseimage` action;

```
jobs:
  build:
    # Main workflow job that builds, tests, etc.

  dispatch:
    needs: build
    strategy:
      matrix:
        repo: ['tiiuae/navigation', 'tiiuae/control_interface', 'tiiuae/peroni']
    runs-on: ubuntu-latest
    steps:
      - name: Repository Dispatch
        uses: peter-evans/repository-dispatch@v2
        with:
          token: ${{ secrets.REPO_ACCESS_TOKEN }}
          repository: ${{ matrix.repo }}
          event-type: base-image-update
```

So the child repos trigger actions again once the event `base-image-update` is emitted from `fog-ros-baseimage` repository. And if the stable branch is implemented and all the workflow is automatized, simple merging of the PR should trigger actions all the nodes that are in the strategy matrix.

https://github.com/marketplace/actions/repository-dispatch
